src/pcrlock/pcrlock.c

```diff
@@ -0,0 +1,5208 @@
+#define PCRLOCK_GPT_PATH                    "/var/lib/pcrlock.d/600-gpt.pcrlock"
+#define PCRLOCK_SECUREBOOT_AUTHORITY_PATH   "/var/lib/pcrlock.d/620-secureboot-authority.pcrlock"
+#define PCRLOCK_KERNEL_CMDLINE_PATH         "/var/lib/pcrlock.d/710-kernel-cmdline.pcrlock"
+#define PCRLOCK_KERNEL_INITRD_PATH          "/var/lib/pcrlock.d/720-kernel-initrd.pcrlock"
+#define PCRLOCK_MACHINE_ID_PATH             "/var/lib/pcrlock.d/820-machine-id.pcrlock"
+#define PCRLOCK_ROOT_FILE_SYSTEM_PATH       "/var/lib/pcrlock.d/830-root-file-system.pcrlock"
+#define PCRLOCK_FILE_SYSTEM_PATH_PREFIX     "/var/lib/pcrlock.d/840-file-system-"
+
+/* The default set of PCRs to lock to */
+#define DEFAULT_PCR_MASK                                     \
+        ((UINT32_C(1) << TPM2_PCR_PLATFORM_CODE) |           \
+         (UINT32_C(1) << TPM2_PCR_PLATFORM_CONFIG) |         \
+         (UINT32_C(1) << TPM2_PCR_EXTERNAL_CODE) |           \
+         (UINT32_C(1) << TPM2_PCR_EXTERNAL_CONFIG) |         \
+         (UINT32_C(1) << TPM2_PCR_BOOT_LOADER_CODE) |        \
+         (UINT32_C(1) << TPM2_PCR_BOOT_LOADER_CONFIG) |      \
+         (UINT32_C(1) << TPM2_PCR_SECURE_BOOT_POLICY) |      \
+         (UINT32_C(1) << TPM2_PCR_KERNEL_BOOT) |             \
+         (UINT32_C(1) << TPM2_PCR_KERNEL_CONFIG) |           \
+         (UINT32_C(1) << TPM2_PCR_SYSEXTS) |                 \
+         (UINT32_C(1) << TPM2_PCR_SHIM_POLICY) |             \
+         (UINT32_C(1) << TPM2_PCR_SYSTEM_IDENTITY))
+
+/* todo:
+ *
+ * pre-merge:
+ *    1. make cryptsetup honour two PCR authnv
+ *    2. include multiple --location= in generated policy
+ *    3. My sure "make-policy" is a quick NOP if policy doesn't change.
+ *    4. add kernel-install plugin that automatically creates UKI .pcrlock file when UKI is installed, and removes it when it is removed again
+ *    5. automatically install PE measurement of sd-boot on "bootctl install"
+ *    6. write generated pcrlock signature files to the ESP as credential, one for each installed OS
+ *    7. pick up generated pcrlock signature file in sd-stub, pass it via initrd to OS
+ *    8. add integration + unit tests
+ *    9. document everything
+ *
+ * post-merge:
+ *    10. pre-calc sysext + kernel cmdline measurements
+ *    11. pre-calc cryptsetup root key measurement
+ *    12. maybe make systemd-repart generate .pcrlock for old and new GPT header in /run?
+ *    13. Add support for more than 8 branches per PCR OR
+ */
+
+typedef struct EventLogRecordBank EventLogRecordBank;
+typedef struct EventLogRecord EventLogRecord;
+typedef struct EventLogRegisterBank EventLogRegisterBank;
+typedef struct EventLogRegister EventLogRegister;
+typedef struct EventLogComponentInstance EventLogComponentInstance;
+typedef struct EventLogComponent EventLogComponent;
+typedef struct EventLog EventLog;
+
+struct EventLogRecordBank {
+        uint16_t algorithm;
+        TPM2B_DIGEST hash;
+        LIST_FIELDS(struct EventLogRecordBank, banks);
+};
```

Comment:
`typedef … EventLogRecordBank`

Reply:
it's already there, just a bunch of lines further up